File: bracket_closure.js
const bracket_closure = s => {
    let req_ending_brackets = [];
    let pointer = 0;

    for (const x of s){
        if (x == '('){
            req_ending_brackets.push(')');
        }
        else if (x == '{'){
            req_ending_brackets.push('}');
        }
        else if (x == '['){
            req_ending_brackets.push(']');
        }
        else if (x == ')' || x == '}' || x == ']'){
            if (req_ending_brackets[pointer] == x){
                req_ending_brackets[pointer] = 0;
                pointer += 1
            }
            else{
                break;
            }
        }
    }
    if (pointer == req_ending_brackets.length){
        return true;
    }
    else{
        return false;
    }
}
// test cases
console.log(bracket_closure('()')); //true
console.log(bracket_closure('()[]{}')); //true
console.log(bracket_closure('(]')); //false
console.log(bracket_closure('(())')); //true
console.log(bracket_closure('(()')); //false

console.log(bracket_closure('[{()}]')); //false

File: bracket_closure.py
def bracket_closure(s):

    req_ending_brackets = []
    pointer = 0

    for x in s:
        if x=='(':
            req_ending_brackets.append(')')
        elif x=='{':
            req_ending_brackets.append('}')
        elif x=='[':
            req_ending_brackets.append(']')

        elif x == ')' or x == '}' or x == ']':
            if req_ending_brackets[pointer] == x:
                req_ending_brackets[pointer] = 0
                pointer += 1
            else:
                break
 
    if (pointer == len(req_ending_brackets)):
        return True
    else:
        return False
    
print(bracket_closure('(()'))